pyteal/ast/safemath.py

```diff
@@ -0,0 +1,159 @@
+from typing import List, Tuple, TYPE_CHECKING
+
+from ..types import TealType, require_type
+from ..errors import TealInputError, TealInternalError, TealCompileError
+from ..ir import TealOp, Op, TealSimpleBlock, TealBlock
+from .expr import Expr
+
+if TYPE_CHECKING:
+    from ..compiler import CompileOptions
+
+
+def multiplyFactors(
+    expr: Expr, factors: List[Expr], options: "CompileOptions"
+) -> Tuple[TealSimpleBlock, TealSimpleBlock]:
+    if len(factors) == 0:
+        raise TealInternalError("Received 0 factors")
+
+    start = TealSimpleBlock([])
+
+    fac0Start, fac0End = factors[0].__teal__(options)
+
+    if len(factors) == 1:
+        # need to use 0 as high word
+        highword = TealSimpleBlock([TealOp(expr, Op.int, 0)])
+
+        start.setNextBlock(highword)
+        highword.setNextBlock(fac0Start)
+
+        end = fac0End
+    else:
+        start.setNextBlock(fac0Start)
+
+        fac1Start, fac1End = factors[1].__teal__(options)
+        fac0End.setNextBlock(fac1Start)
+
+        multiplyFirst2 = TealSimpleBlock([TealOp(expr, Op.mulw)])
+        fac1End.setNextBlock(multiplyFirst2)
+
+        end = multiplyFirst2
+        for factor in factors[2:]:
+            facXStart, facXEnd = factor.__teal__(options)
+            end.setNextBlock(facXStart)
+
+            # stack is [..., A, B, C], where C is current factor
+            # need to pop all A,B,C from stack and push X,Y, where X and Y are:
+            #       X * 2**64 + Y = (A * 2**64 + B) * C
+            # <=>   X * 2**64 + Y = A * C * 2**64 + B * C
+            # <=>   X = A * C + highword(B * C)
+            #       Y = lowword(B * C)
+            multiply = TealSimpleBlock(
+                [
+                    TealOp(expr, Op.uncover, 2),  # stack: [..., B, C, A]
+                    TealOp(expr, Op.dig, 1),  # stack: [..., B, C, A, C]
+                    TealOp(expr, Op.mul),  # stack: [..., B, C, A*C]
+                    TealOp(expr, Op.cover, 2),  # stack: [..., A*C, B, C]
+                    TealOp(
+                        expr, Op.mulw
+                    ),  # stack: [..., A*C, highword(B*C), lowword(B*C)]
+                    TealOp(
+                        expr, Op.cover, 2
+                    ),  # stack: [..., lowword(B*C), A*C, highword(B*C)]
+                    TealOp(
+                        expr, Op.add
+                    ),  # stack: [..., lowword(B*C), A*C+highword(B*C)]
+                    TealOp(
+                        expr, Op.swap
+                    ),  # stack: [..., A*C+highword(B*C), lowword(B*C)]
+                ]
+            )
+
+            facXEnd.setNextBlock(multiply)
+            end = multiply
+
+    return start, end
+
+
+class SafeRatio(Expr):
+    """A class used to calculate expressions of the form :code:`(N_1 * N_2 * N_3 * ...) / (D_1 * D_2 * D_3 * ...)`
+
+    Use this class if all inputs to the expression are uint64s, the output fits in a uint64, and all
+    intermediate values fit in a uint128.
```

Comment:
The reason to allow > 2 numerators is because I believe an interface that only allows `A*B/C` is pretty restrictive. If we only exposed that interface and you needed to have 3 numerators, then you would have to do `(A_1*A_2)*B/C`, which is strictly worse since `A_1*A_2` has to fit in a uint64, instead of `A_1*A_2*B` having to fit in a uint128.

Plus, due to integer rounding behavior during division, I'm not sure it would be possible to implement the optimization that you described.

Reply:
You could do A_1; A_2; mulw; C; divmodw; B mul. (taking some liberties with stack management in there). That would keep intermediates smaller.  But I don't know how badly rounding would compound imprecision.

Anyway, since this does A*B/C efficiently, I don't see any reason to complain much about it doing more if you can ask it to do more. Though I guess I'd argue that the docs should be clearer about the order of the steps, so it's clearer what is meant by "intermediate values".